crates/distribution-types/src/prioritized_distribution.rs

```diff
@@ -11,128 +11,105 @@ pub struct PrioritizedDist(Box<PrioritizedDistInner>);
 /// [`PrioritizedDist`] is boxed because [`Dist`] is large.
 #[derive(Debug, Default, Clone)]
 struct PrioritizedDistInner {
-    /// An arbitrary source distribution for the package version.
-    source: Option<DistMetadata>,
-    /// The most-compatible wheel distribution for the package version.
-    wheel: Option<(DistMetadata, WheelCompatibility)>,
+    /// TODO(zanieb)
+    source: Option<(Dist, SourceDistCompatibility)>,
+    ///  TODO(zanieb)
+    wheel: Option<(Dist, WheelCompatibility)>,
     /// The hashes for each distribution.
     hashes: Vec<Hashes>,
-    /// If exclude newer filtered files from this distribution
-    exclude_newer: bool,
 }
 
 /// A distribution that can be used for both resolution and installation.
 #[derive(Debug, Clone)]
 pub enum CompatibleDist<'a> {
     /// The distribution should be resolved and installed using a source distribution.
-    SourceDist(&'a DistMetadata),
+    SourceDist(&'a Dist),
     /// The distribution should be resolved and installed using a wheel distribution.
-    CompatibleWheel(&'a DistMetadata, TagPriority),
+    CompatibleWheel(&'a Dist, TagPriority),
     /// The distribution should be resolved using an incompatible wheel distribution, but
     /// installed using a source distribution.
     IncompatibleWheel {
-        source_dist: &'a DistMetadata,
-        wheel: &'a DistMetadata,
+        source_dist: &'a Dist,
+        wheel: &'a Dist,
     },
 }
 
 #[derive(Debug, PartialEq, Eq, Clone)]
+pub enum IncompatibleDist {
+    /// An incompatible wheel is available.
+    Wheel(IncompatibleWheel),
+    /// An incompatible source distribution is available.
+    Source(IncompatibleSource),
+    /// No distributions are available
+    Unavailable,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
 pub enum WheelCompatibility {
     Incompatible(IncompatibleWheel),
     Compatible(TagPriority),
 }
 
-#[derive(Debug, PartialEq, Eq, Ord, PartialOrd, Clone)]
+#[derive(Debug, PartialEq, Eq, Clone)]
 pub enum IncompatibleWheel {
+    ExcludeNewer(Option<i64>),
     Tag(IncompatibleTag),
-    RequiresPython,
+    RequiresPython(VersionSpecifiers),
+    Yanked(Yanked),
     NoBinary,
 }
 
-/// A [`Dist`] and metadata about it required for downstream filtering.
-#[derive(Debug, Clone)]
-pub struct DistMetadata {
-    /// The distribution.
-    pub dist: Dist,
-    /// The version of Python required by the distribution.
-    pub requires_python: Option<VersionSpecifiers>,
-    /// If the distribution file is yanked.
-    pub yanked: Yanked,
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum SourceDistCompatibility {
+    Incompatible(IncompatibleSource),
+    Compatible,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum IncompatibleSource {
+    ExcludeNewer(Option<i64>),
+    RequiresPython(VersionSpecifiers),
+    Yanked(Yanked),
```

Comment:
These three overlap with variants in `IncompatibleWheel`. Should we combine them into a structure like:

```
Incompatible
    ExcludeNewer
    RequiresPython
    Yanked
    Wheel
        NoBinary
        Tag
    SourceDist
        NoBuild
```